pybamm/simulation.py

```diff
@@ -16,84 +16,109 @@ class Simulation(object):
+        self,
+        model,
+        parameter_values=None,
+        discretisation=None,
+        solver=None,
+        name="unnamed",
+    ):
+        # Read defaults from model
+        if parameter_values is None:
+            parameter_values = model.default_parameter_values
+        if discretisation is None:
+            discretisation = model.default_discretisation
         if solver is None:
-            solver = pybamm.Solver()
+            solver = model.default_solver
 
         # Assign attributes
         self.model = model
-        self.param = param
-        self.mesh = mesh
+        self.parameter_values = parameter_values
+        self.discretisation = discretisation
         self.solver = solver
         self.name = name
 
-        # Initialise simulation to prepare for solving
-        # Set mesh dependent parameters
-        self.param.set_mesh(self.mesh)
-
-        # Create operators from solver
-        self.operators = self.solver.operators(self.mesh)
-
-        # Assign param, operators and mesh as model attributes
-        self.model.set_simulation(self.param, self.operators, self.mesh)
+        # Default evaluation time
+        self.default_t_eval = np.linspace(0, 1)
 
     def __str__(self):
         return self.name
 
-    def run(self, use_force=False):
-        """
-        Run the simulation.
-
-        Parameters
-        ----------
-        use_force : boolean, optional
-            If False (default), use a stored solution (if it exists). Otherwise, run the
-            model.
-        """
-        # if not use_force and os.path.isfile(self.filename):
-        #     self.load()
-        # else:
-        self.vars = self.solver.get_simulation_vars(self)
-        # self.save()
-
-    def average(self):
-        """Average simulation variable attributes over the relevant (sub)domain."""
-        self.vars.average()
-
-    @property
-    def filename_io(self):
-        path = os.path.join("out", "simulations")
-        filename = "model_{}_solver_{}".format(
-            pickle.dumps(self.model), pickle.dumps(self.solver)
-        )
-        return os.path.join(path, filename)
+    def set_parameters(self):
+        self.parameter_values.process_model(self.model)
+
+    def discretise(self):
+        self.discretisation.process_model(self.model)
+
+    def solve(self, t_eval=None):
+        if t_eval is None:
+            t_eval = self.default_t_eval
+        self.solver.solve(self.model, t_eval)
+
+    def run(self, t_eval=None):
+        self.set_parameters()
+        self.discretise()
+        self.solve(t_eval)
 
     def load(self):
-        """Load saved simulation if it exists."""
-        with open(self.filename_io, "rb") as input:
-            return pickle.load(input)
+        raise NotImplementedError
 
     def save(self):
-        """Save simulation."""
         raise NotImplementedError
 
     def plot(self):
         raise NotImplementedError
+
+
+# QUESTION: is there a way to test this?
```

Comment:
you could use `result = os.system("python simulation.py")` and check the result? Not sure that would work with the coverage, but it would at least run it